components/suggest/src/schema.rs

```diff
@@ -195,6 +196,56 @@ pub fn clear_database(db: &Connection) -> rusqlite::Result<()> {
     )
 }
 
+fn define_functions(c: &Connection) -> rusqlite::Result<()> {
+    c.create_scalar_function(
+        "hash",
+        1,
+        FunctionFlags::SQLITE_UTF8 | FunctionFlags::SQLITE_DETERMINISTIC,
+        sql_fns::hash,
+    )?;
+    Ok(())
+}
+
+mod sql_fns {
+    use md5::{Digest, Md5};
```

Comment:
With my Places peer hat on 🧙🏼‍♀️ I should call out that the Desktop security engineering team has been working on [removing deprecated hashes from Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1871103#c1), including MD5. Places recently [switched to using SHA-256](http://bugzilla.mozilla.org/show_bug.cgi?id=1871103) for cases where it previously used MD5.

Even though we have a known and trusted set of URLs here, I think introducing a new use of a deprecated hash function is still going to raise a few eyebrows when we vendor.

Reply:
It looks like the average length of `suggestions.url` is ~64 bytes. SHA-256 digests are 32 bytes, so it looks like it's still a bit more space-efficient on average to use a hash instead of the full URL.

That said, though, I think storing the full URL should also be OK, especially if it turns out we need it because of the `%%YYYYMMDDHH%` replacement parameters.

In general, I think a hash is going to be more efficient than a long text primary key or index when there are a lot of rows, as in Places (tens of thousands). I don't think we expect dismissed suggestions to be anywhere close to that size, though? For dozens or hundreds of rows, storing the URL directly is A-OK.

Reply:
Between moving to SHA-256 and using the full URL, I think we should just use the full one.  As you point out the space difference is small and we shouldn't expect many URLs to be stored in the table.  @ncloudioj how do you feel about it?

Reply:
Thanks for the heads up, Lina!

I am still leaning towards hashing those dismissed URLs to be consistent with what we are doing for other similar cases in the browser (e.g. we MD5 hash the dismissed URLs for New Tab; we SHA1 hash dismissed URLs for Firefox Suggest Desktop). We can pick SHA-256 to stay away from those insecure hashers. If we don't store the whole digest, we could grab a smaller portion of it (e.g. 16 bytes) which should offer enough entropy given the relatively smaller input size.

Does that make sense?

Reply:
> we could grab a smaller portion of it (e.g. 16 bytes) which should offer enough entropy given the relatively smaller input size.

That was my other question 😅—it looks like you're only keeping the first 8 bytes for the primary key, so any collisions will exclude all the colliding URLs. Is that right? That still sounds OK for our case, but definitely subtle.

Even if we used the full 32-byte hash, though, it's still going to be shorter than any URLs—the shortest URL we have is `https://en.wikipedia.org/wiki/LVM3`, which is 34 bytes.

For comparison, Places uses a [non-cryptographic 48-bit hash](https://github.com/mozilla/application-services/blob/9054db4bb5031881550ceab3448665ef6499a706/components/places/src/hash.rs), with a [non-unique index](https://github.com/mozilla/application-services/blob/9054db4bb5031881550ceab3448665ef6499a706/components/places/sql/create_shared_schema.sql#L38)—so if there are any collisions, it'll still compare URLs, but the index reduces the search space from tens of thousands to a handful. But that scheme pays off for larger tables—for smaller tables like ours, it can end up increasing the on-disk size.

Do you happen to know why we store hashes on Desktop, @ncloudioj and @0c0w3? Is it mostly for size benefits, or are there other reasons like privacy?

Reply:
To my understanding, we hash the dismissed URLs for both benefits: 1), we always store a fixed size hash digest regardless of the size of the original URL; 2), as both cases persist the dismissed URLs in preferences, the hashed digests don't expose anything in plain as they can be viewed / accessed by unauthorized individuals or addons.

I am not super worried about the size actually, we can store the entire 32-byte hash as is in the db. Alternatively, there are a whole lot of other non-cryptographic hashers available such as [smhasher](https://github.com/rurban/smhasher) with Rust implementations, also the one Lina mentioned above at our disposal.

I don't want to block us for this though, happy to move on with the option of storing raw URLs :)

Reply:
How about we start with raw URLs and consider adding the hashes as a performance/privacy improvement later (although  I have some questions on how much of a privacy gain this is).

Reply:
Yup, let's revisit this later.

> although I have some questions on how much of a privacy gain this is.

Curious about your questions, can you share?

Reply:
The suggestion list is public and also fairly small.  So shouldn't an attacker be able to create a map of hashed URLs -> full URLs.  This won't be perfect if there's a hash collision, but it seems like this would get around most of the privacy protections.  Maybe adding a salt component would fix this though?  Not sure on that one.

Reply:
Ah, I see. Your point makes total sense in this context.

I think for all those dismissal scenarios, we were not attempting to apply standard mechanisms (salting & encrypting) to secure the data or preserve users' privacy in the best effort. Rather, we meant to convert the data in a way such that the processed data is not obvious to read. Imagine one dismissed a URL from the New Tab page because they wanted to "hide" it, storing it in plain in another user visible place could cause some surprises & awkwardness later. Other benefits (such as smaller footprint & easier to index/query) are a plus so I think that's why decided to hash the raw URLs.

Reply:
> How about we start with raw URLs

That sounds great to me!